workers/src/verifier.rs

```diff
@@ -0,0 +1,267 @@
+use crate::config::DVNConfig;
+use alloy::eips::BlockNumberOrTag;
+use alloy::network::Ethereum;
+use alloy::primitives::B256;
+use alloy::providers::{Provider, ProviderBuilder, ReqwestProvider};
+use reqwest::{Client, ClientBuilder, Url};
+use serde::{Deserialize, Serialize};
+use std::time::Duration;
+use tracing::error;
+
+#[derive(Serialize, Deserialize, std::fmt::Debug)]
+pub(crate) struct ResponseWrapper {
+    pub message: Message,
+}
+
+#[derive(Serialize, Deserialize, std::fmt::Debug)]
+pub(crate) struct Message {
+    #[serde(rename = "RollupId")]
+    rollup_id: u32,
+    #[serde(rename = "BlockHeight")]
+    block_height: u64,
+    #[serde(rename = "Timestamp")]
+    timestamp: u64,
+    #[serde(rename = "NearDaTransactionId")]
+    near_da_transaction_id: Vec<u8>,
+    #[serde(rename = "NearDaCommitment")]
+    near_da_commitment: Vec<u8>,
+    #[serde(rename = "StateRoot")]
+    pub(crate) state_root: Vec<u8>,
+}
+
+// TODO: Generify in a future for other networks, like Solana.
+/// Verifies state roots for the Decentralized Verification Network (DVN).
+///
+/// This verifier implements the V1 verification algorithm which:
+/// 1. Fetches the state root from the aggregator
+/// 2. Retrieves the block state root using the chain RPC API
+/// 3. Compares the two state roots to determine message validity
+pub struct NFFLVerifier {
+    http_client: Client,
+    eth_l2_provider: ReqwestProvider<Ethereum>,
+    aggregator_http_address: String,
+    network_id: String,
+}
+
+impl NFFLVerifier {
+    pub(crate) async fn new(agg_url: &str, eth_l2_url: &str, network_id: u64) -> eyre::Result<NFFLVerifier> {
+        let client = ClientBuilder::new().build()?;
+        let url: Url = eth_l2_url.parse()?;
+        let provider = ProviderBuilder::new().on_http(url);
+        let agg_http_addr = format!("{}/state-root", agg_url);
```

Comment:
Note: this might need to change according to `GetStateRootUpdateAggregation` discussion